gobblin-runtime/src/main/java/gobblin/runtime/Fork.java

```diff
@@ -70,9 +72,26 @@
   private final RowLevelPolicyCheckResults rowLevelPolicyCheckingResult;
   private final DataWriter<Object> writer;
 
+  // This is used to signal the parent task of the completion of this fork
+  private final CountDownLatch countDownLatch;
+
+  // A bounded blocking queue in between the parent task and this fork
+  private final BoundedBlockingRecordQueue<Object> recordQueue;
+
   private final Closer closer = Closer.create();
 
-  public Fork(TaskContext taskContext, TaskState taskState, Object schema, int branches, int index)
+  // This is used by the parent task to signal that it has done pulling records and this fork
+  // should not expect any new incoming data records
+  private volatile boolean parentTaskDone = false;
+
+  // This tells if this fork is done
+  private volatile boolean done = false;
```

Comment:
Can we give a better name here, instead of "done"? It is more like a indicator showing if the current fork is running or not. Maybe call it "alive"? It is a little confusing to me to have both "done" and "succeeded" here for two purposes.